[SYCL][FPGA] Allow use_stall_enable_clusters attribute to kernel

smanna12 opened the pull request:
This patch collects and applies the FPGA attribute intel::use_stall_enable_clusters to the callers/SYCL kernel if directly applied through functors/lambda function.

The attribute has to be applicable to all functions, which can include the SYCL kernels and ​must not be propagated up to the caller/SYCL kernel when called from a function.

This patch fixes FPGA emulator bug that was introduced on https://github.com/intel/llvm/pull/3900.

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

smanna12 commented:
I talked to @mendell27. The attribute ``intel::use_stall_enable_clusters`` need to be applied to the callers/SYCL kernel if directly applied through functors/lambda function.

erichkeane reviewed (commented):
Can you please include a test that shows that your comment in AttrDocs.td is true?

smanna12 commented:
> Can you please include a test that shows that your comment in AttrDocs.td is true?

@erichkeane, Thanks for the review. I have already a test for this for both CodeGen and SemaSYCL:
```[[intel::use_stall_enable_clusters]] void test() {}

SemaSYCL:
// Test attribute is not propagated to the kernel.
    // CHECK-LABEL: FunctionDecl {{.*}}test_kernel3
    // CHECK-NOT:   SYCLIntelUseStallEnableClustersAttr {{.*}}
    h.single_task<class test_kernel3>(
        []() { test(); });

CodeGenSYCL:
// CHECK: define {{.*}}spir_kernel void @{{.*}}test_kernel3() #0 !kernel_arg_buffer_location ![[NUM5]]
    // CHECK: define {{.*}}spir_func void @_Z4testv() #3 !stall_enable ![[NUM4]]
    h.single_task<class test_kernel3>(
        []() { test(); });
// CHECK: ![[NUM4]] = !{i32 1}
// CHECK: ![[NUM5]] = !{}

erichkeane commented:
> 
> 
> > Can you please include a test that shows that your comment in AttrDocs.td is true?
> 
> @erichkeane, Thanks for the review. I have already a test for this for both CodeGen and SemaSYCL:
> 
> ```
> [[intel::use_stall_enable_clusters]] void test() {}
> 
> SemaSYCL:
> // Test attribute is not propagated to the kernel.
>     // CHECK-LABEL: FunctionDecl {{.*}}test_kernel3
>     // CHECK-NOT:   SYCLIntelUseStallEnableClustersAttr {{.*}}
>     h.single_task<class test_kernel3>(
>         []() { test(); });
> 
> CodeGenSYCL:
> // CHECK: define {{.*}}spir_kernel void @{{.*}}test_kernel3() #0 !kernel_arg_buffer_location ![[NUM5]]
>     // CHECK: define {{.*}}spir_func void @_Z4testv() #3 !stall_enable ![[NUM4]]
>     h.single_task<class test_kernel3>(
>         []() { test(); });
> // CHECK: ![[NUM4]] = !{i32 1}
> // CHECK: ![[NUM5]] = !{}
> ```

Based on the diff, you removed at least the sema one!

smanna12 commented:
> > > Can you please include a test that shows that your comment in AttrDocs.td is true?
> > 
> > 
> > @erichkeane, Thanks for the review. I have already a test for this for both CodeGen and SemaSYCL:
> > ```
> > [[intel::use_stall_enable_clusters]] void test() {}
> > 
> > SemaSYCL:
> > // Test attribute is not propagated to the kernel.
> >     // CHECK-LABEL: FunctionDecl {{.*}}test_kernel3
> >     // CHECK-NOT:   SYCLIntelUseStallEnableClustersAttr {{.*}}
> >     h.single_task<class test_kernel3>(
> >         []() { test(); });
> > 
> > CodeGenSYCL:
> > // CHECK: define {{.*}}spir_kernel void @{{.*}}test_kernel3() #0 !kernel_arg_buffer_location ![[NUM5]]
> >     // CHECK: define {{.*}}spir_func void @_Z4testv() #3 !stall_enable ![[NUM4]]
> >     h.single_task<class test_kernel3>(
> >         []() { test(); });
> > // CHECK: ![[NUM4]] = !{i32 1}
> > // CHECK: ![[NUM5]] = !{}
> > ```
> 
> > > Based on the diff, you removed at least the sema one!
> 
> The test is not showing on the diff because it is an existing test in sema. I did not add any new one for the current diff. So Sema test is still there.
> 
> > > Also, the codegen one is testing that it IS there, what about when it is ignored?
> 
> ```
> [[intel::use_stall_enable_clusters]] void test() {}
> 
> No, it is ignored. The attribute does not get propagated to the kernel below, i.e. spir_kernel.
> // CHECK: define {{.*}}spir_kernel void @{{.*}}test_kernel3() #0 !kernel_arg_buffer_location ![[NUM5]]
> 
> Here the metadata is attached to the functionDecl that the attribute is applied to test().
>  // CHECK: define {{.*}}spir_func void @_Z4testv() #3 !stall_enable ![[NUM4]]
> 
>  h.single_task<class test_kernel3>(
>       []() { test(); });
> // CHECK: ![[NUM5]] = !{}
> // CHECK: ![[NUM4]] = !{i32 1}
> ```

@erichkeane, i have updated tests. Added comments and showed that If ``intel::use_stall_enable_clusters`` is applied to a function called from a device kernel, the attribute is ignored and it is not propagated to the kernel.

elizabethandrews commented on clang/test/CodeGenSYCL/stall_enable_device.cpp:
```diff
@@ -2,20 +2,21 @@
 // Tests for IR of Intel FPGA [[intel::use_stall_enable_clusters]] function attribute on Device.
 // The metadata to be attached to the functionDecl that the attribute is applied to.
-// The attributes do not get propagated to kernel metadata i.e. spir_kernel.
+// The attributes get propagated to the kernel metadata i.e. spir_kernel if directly applied
```
I don't think this comment is required. This is standard behavior for SYCL attributes, which is why I was surprised the behavior was different for this attribute in https://github.com/intel/llvm/pull/3900. You can probably just say - `// Tests for IR of Intel FPGA [[intel::use_stall_enable_clusters]] function attribute on Device`

elizabethandrews commented on clang/test/CodeGenSYCL/stall_enable_device.cpp:
```diff
@@ -25,30 +26,49 @@ class Foo {
   [[intel::use_stall_enable_clusters]] void operator()() const {}
 };
 
+class Functor {
```
Why do you need this ? `Foo` is the same thing

AaronBallman reviewed (approved):
LGTM

elizabethandrews commented on clang/test/CodeGenSYCL/stall_enable_device.cpp:
```diff
@@ -25,30 +26,49 @@ class Foo {
+
+    // CHECK: define {{.*}}spir_kernel void @{{.*}}test_kernel6() {{.*}} !stall_enable ![[NUM4]]
+    // CHECK: define {{.*}}spir_func void @{{.*}}Functor{{.*}}(%class._ZTS7Functor.Functor addrspace(4)* align 1 dereferenceable_or_null(1) %this) #3 comdat align 2 !stall_enable ![[NUM4]]
+    Functor f1;
```
Isn't this testing the same thing as L43?

elizabethandrews commented on clang/test/SemaSYCL/stall_enable_device.cpp:
```diff
@@ -24,7 +25,7 @@ struct FuncObj {
   // CHECK: CXXRecordDecl{{.*}}implicit struct FuncObj
   // CHECK-NEXT: CXXMethodDecl{{.*}}used operator() 'void () const'
-  // CHECK-NEXT-NEXT:SYCLIntelUseStallEnableClustersAttr
+  // CHECK-NEXT-NEXT:SYCLIntelUseStallEnableClustersAttr{{.*}}
```
I don't think you need to add `{{.*}}` in the end

elizabethandrews commented on clang/test/SemaSYCL/stall_enable_device.cpp:
```diff
@@ -33,28 +34,46 @@ void test3() {
   // CHECK: FunctionDecl{{.*}}test3
   // CHECK: LambdaExpr
-  // CHECK: SYCLIntelUseStallEnableClustersAttr
+  // CHECK: SYCLIntelUseStallEnableClustersAttr{{.*}}
```
Here as well

smanna12 commented:
Thanks @AaronBallman for the reviews.

AaronBallman commented:
Sorry for merging before the review from @elizabethandrews arrived! Please address those concerns in a follow-up PR.

smanna12 commented:
Thanks @elizabethandrews  for the reviews. I will address all your comments in a separate patch (they are mostly NFC changes).

elizabethandrews commented on clang/test/SemaSYCL/stall_enable_device.cpp:
```diff
@@ -33,28 +34,46 @@ void test3() {
+    // Test attribute is presented on LambdaExpr called by kernel.
     // CHECK-LABEL: FunctionDecl {{.*}}test_kernel2
-    // CHECK-NOT:   SYCLIntelUseStallEnableClustersAttr {{.*}}
+    // CHECK-:      SYCLIntelUseStallEnableClustersAttr {{.*}}
```
Sorry missed this. Extra `-` after `CHECK`

smanna12 commented:
> Thanks @elizabethandrews for the reviews. I will address all your comments in a follow-up patch (they are mostly NFC changes).

@elizabethandrews all your review comments and created PR is here: https://github.com/intel/llvm/pull/4060
